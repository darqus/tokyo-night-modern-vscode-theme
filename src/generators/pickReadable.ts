import type { InterfacePalette } from '../types/theme'
import { getContrastRatioAware } from '../core/contrast'
import {
  type SurfaceType,
  type BaseSurface,
  type ContrastThresholds,
  resolveSurface,
  DEFAULT_CONTRAST_THRESHOLDS,
} from '../core/surfaces'

// Совместимость: старый тип для существующего кода
export type SurfaceKey = BaseSurface

type Thresholds = {
  primaryMin?: number
  mutedMin?: number
  subtleMin?: number
}

const DEFAULTS: Required<Thresholds> = {
  primaryMin: 4.5,
  mutedMin: 3.0,
  subtleMin: 2.5,
}

/**
 * Улучшенная версия pickReadable с поддержкой новой системы поверхностей
 */
export function pickReadableAdvanced(
  ip: InterfacePalette,
  surface: SurfaceType,
  preferred: 'muted' | 'subtle' = 'muted',
  customThresholds?: Partial<ContrastThresholds>
): string {
  const baseSurface = resolveSurface(surface)
  const thresholds = { ...DEFAULT_CONTRAST_THRESHOLDS, ...customThresholds }

  const bg = getSurfaceBackground(ip, baseSurface)
  const roles = ip.textOn[baseSurface]

  if (preferred === 'subtle') {
    const rSubtle = getContrastRatioAware(roles.subtle, bg as any)
    if (rSubtle >= thresholds.subtleMin) return roles.subtle

    // Fallback to muted
    const rMuted = getContrastRatioAware(roles.muted, bg as any)
    if (rMuted >= thresholds.mutedMin) return roles.muted

    return roles.primary
  }

  // preferred === 'muted'
  const rMuted = getContrastRatioAware(roles.muted, bg as any)
  if (rMuted >= thresholds.mutedMin) return roles.muted

  return roles.primary
}

/**
 * Получить фон для базовой поверхности
 */
function getSurfaceBackground(
  ip: InterfacePalette,
  surface: BaseSurface
): string {
  switch (surface) {
    case 'base':
      return ip.bg.base
    case 'elevated':
      return ip.bg.elevated
    case 'overlay':
      return ip.bg.overlay
    default:
      return ip.bg.base
  }
}

/**
 * Выбор читаемой роли текста для заданной поверхности.
 * Предпочитаем muted/subtle, но при нехватке контраста возвращаем primary.
 *
 * @deprecated Используйте pickReadableAdvanced для новых возможностей
 */
export function pickReadable(
  ip: InterfacePalette,
  surface: SurfaceKey,
  preferred: 'muted' | 'subtle' = 'muted',
  thresholds: Thresholds = {}
): string {
  const t = { ...DEFAULTS, ...thresholds }
  const bg = getSurfaceBackground(ip, surface)
  const roles = ip.textOn[surface]

  if (preferred === 'subtle') {
    const rSubtle = getContrastRatioAware(roles.subtle, bg as any)
    if (rSubtle >= t.subtleMin) return roles.subtle
    // если subtle не проходит — пробуем muted
    const rMuted = getContrastRatioAware(roles.muted, bg as any)
    if (rMuted >= t.mutedMin) return roles.muted
    return roles.primary
  }

  // preferred === 'muted'
  const rMuted = getContrastRatioAware(roles.muted, bg as any)
  if (rMuted >= t.mutedMin) return roles.muted
  return roles.primary
}

/**
 * Выбор оптимального цвета текста с учетом accessibility требований
 */
export function pickAccessibleText(
  ip: InterfacePalette,
  surface: SurfaceType,
  wcagLevel: 'AA' | 'AAA' = 'AA'
): string {
  const baseSurface = resolveSurface(surface)
  const bg = getSurfaceBackground(ip, baseSurface)
  const roles = ip.textOn[baseSurface]

  const targetContrast = wcagLevel === 'AAA' ? 7.0 : 4.5

  // Проверяем в порядке приоритета
  const candidates = [
    { role: 'primary', color: roles.primary },
    { role: 'muted', color: roles.muted },
    { role: 'subtle', color: roles.subtle },
    { role: 'inactive', color: roles.inactive },
  ]

  for (const candidate of candidates) {
    const contrast = getContrastRatioAware(candidate.color, bg as any)
    if (contrast >= targetContrast) {
      return candidate.color
    }
  }

  // Fallback к primary если ничего не прошло
  return roles.primary
}

/**
 * Получить все доступные роли текста для поверхности с их контрастом
 */
export function getTextRolesWithContrast(
  ip: InterfacePalette,
  surface: SurfaceType
): Array<{ role: string; color: string; contrast: number }> {
  const baseSurface = resolveSurface(surface)
  const bg = getSurfaceBackground(ip, baseSurface)
  const roles = ip.textOn[baseSurface]

  return [
    {
      role: 'primary',
      color: roles.primary,
      contrast: getContrastRatioAware(roles.primary, bg as any),
    },
    {
      role: 'muted',
      color: roles.muted,
      contrast: getContrastRatioAware(roles.muted, bg as any),
    },
    {
      role: 'subtle',
      color: roles.subtle,
      contrast: getContrastRatioAware(roles.subtle, bg as any),
    },
    {
      role: 'inactive',
      color: roles.inactive,
      contrast: getContrastRatioAware(roles.inactive, bg as any),
    },
  ]
}
