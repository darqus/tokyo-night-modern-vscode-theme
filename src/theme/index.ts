export * from './types';
export * from './palette';
export * from './utils';
export * from './generator';
