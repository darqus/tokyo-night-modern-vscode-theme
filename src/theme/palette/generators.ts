/**
 * Функции генерации цветов на основе базовой палитры
 * Реализует механизмы изменения яркости и прозрачности
 */

/**
 * Преобразует HEX цвет в HSL
 */
function hexToHsl(hex: string): { h: number; s: number; l: number } {
  // Удаление символа # и разбиение на компоненты
  const cleanHex = hex.replace('#', '')
  const r = parseInt(cleanHex.substring(0, 2), 16) / 255
  const g = parseInt(cleanHex.substring(2, 4), 16) / 255
  const b = parseInt(cleanHex.substring(4, 6), 16) / 255

  const max = Math.max(r, g, b)
  const min = Math.min(r, g, b)
  let h = 0
  let s = 0
  const l = (max + min) / 2

  if (max === min) {
    h = s = 0 // achromatic
  } else {
    const d = max - min
    s = l > 0.5 ? d / (2 - max - min) : d / (max + min)

    switch (max) {
      case r:
        h = (g - b) / d + (g < b ? 6 : 0)
        break
      case g:
        h = (b - r) / d + 2
        break
      case b:
        h = (r - g) / d + 4
        break
    }

    h /= 6
  }

  return {
    h: Math.round(h * 360),
    s: Math.round(s * 100),
    l: Math.round(l * 100),
  }
}

/**
 * Преобразует HSL цвет в HEX
 */
function hslToHex(h: number, s: number, l: number): string {
  h = h / 360
  s = s / 100
  l = l / 100

  const hue2rgb = (p: number, q: number, t: number): number => {
    if (t < 0) t += 1
    if (t > 1) t -= 1
    if (t < 1 / 6) return p + (q - p) * 6 * t
    if (t < 1 / 2) return q
    if (t < 2 / 3) return p + (q - p) * (2 / 3 - t) * 6
    return p
  }

  let r: number, g: number, b: number

  if (s === 0) {
    r = g = b = l // achromatic
  } else {
    const q = l < 0.5 ? l * (1 + s) : l + s - l * s
    const p = 2 * l - q
    r = hue2rgb(p, q, h + 1 / 3)
    g = hue2rgb(p, q, h)
    b = hue2rgb(p, q, h - 1 / 3)
  }

  const toHex = (c: number): string => {
    const hex = Math.round(c * 255).toString(16)
    return hex.length === 1 ? '0' + hex : hex
  }

  return `#${toHex(r)}${toHex(g)}${toHex(b)}`
}

/**
 * Изменяет яркость цвета
 * @param color - исходный цвет в формате HEX
 * @param amount - величина изменения яркости от -1 до 1 (0.1 = +10% яркости)
 */
export function adjustBrightness(color: string, amount: number): string {
  const hsl = hexToHsl(color)
  let newL = hsl.l + amount * 100
  newL = Math.max(0, Math.min(100, newL)) // Ограничение от 0 до 100
  return hslToHex(hsl.h, hsl.s, newL)
}

/**
 * Делает цвет светлее
 * @param color - исходный цвет в формате HEX
 * @param amount - величина осветления от 0 до 1 (0.1 = +10% яркости)
 */
export function lighten(color: string, amount: number): string {
  return adjustBrightness(color, Math.abs(amount))
}

/**
 * Делает цвет темнее
 * @param color - исходный цвет в формате HEX
 * @param amount - величина затемнения от 0 до 1 (0.1 = -10% яркости)
 */
export function darken(color: string, amount: number): string {
  return adjustBrightness(color, -Math.abs(amount))
}

/**
 * Добавляет прозрачность к цвету
 * @param color - исходный цвет в формате HEX
 * @param alpha - значение прозрачности от 0 до 1 (0.5 = 50% прозрачности)
 */
export function withAlpha(color: string, alpha: number): string {
  const cleanHex = color.replace('#', '')
  const alphaHex = Math.round(alpha * 255)
    .toString(16)
    .padStart(2, '0')
  return `#${cleanHex}${alphaHex}`
}

/**
 * Интерфейс для системы прозрачности
 */
export interface AlphaSystem {
  subtle: string // 20% (33)
  hover: string // 40% (66)
  active: string // 60% (99)
  semiOpaque: string // 80% (cc)
}

/**
 * Создает систему прозрачности на основе значений в шестнадцатеричном формате
 */
export function createAlphaSystem(): AlphaSystem {
  return {
    subtle: '33', // 20% прозрачности
    hover: '66', // 40% прозрачности
    active: '99', // 60% прозрачности
    semiOpaque: 'cc', // 80% прозрачности
  }
}

/**
 * Возвращает цвет с заданным уровнем прозрачности из системы
 * @param color - исходный цвет в формате HEX
 * @param alphaKey - ключ уровня прозрачности
 * @param alphaSystem - система прозрачности
 */
export function getColorWithAlpha(
  color: string,
  alphaKey: keyof AlphaSystem,
  alphaSystem: AlphaSystem = createAlphaSystem()
): string {
  const hex = color.replace('#', '')
  return `#${hex}${alphaSystem[alphaKey]}`
}
