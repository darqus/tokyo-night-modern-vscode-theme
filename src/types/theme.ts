import type { Hex } from './palette'

/**
 * Основные типы для улучшенной архитектуры темы
 */

export interface ThemeVariant {
  id: string
  name: string
  displayName: string
  baseConfig: ThemeConfig
}

export interface ThemeConfig {
  contrast: 'low' | 'normal' | 'high'
  saturation: 'muted' | 'normal' | 'vibrant'
  brightness: 'darker' | 'normal' | 'lighter'
}

export interface ComponentStates {
  default: ComponentStyle
  hover?: ComponentStyle
  active?: ComponentStyle
  focus?: ComponentStyle
  disabled?: ComponentStyle
}

export interface ComponentStyle {
  background?: Hex | 'transparent'
  foreground?: Hex
  border?: Hex | 'transparent'
  shadow?: string
}

export interface UIComponent {
  name: string
  states: ComponentStates
  variants?: ComponentVariants
}

export interface ComponentVariants {
  [key: string]: ComponentStates
}

export interface ColorScale {
  50: Hex // Самый светлый
  100: Hex
  200: Hex
  300: Hex
  400: Hex
  500: Hex // Базовый
  600: Hex
  700: Hex
  800: Hex
  900: Hex // Самый темный
}

export interface ValidationRule {
  name: string
  description: string
  validate: (theme: ThemeData) => ValidationResult
}

export interface ValidationResult {
  passed: boolean
  issues: ValidationIssue[]
}

export interface ValidationIssue {
  severity: 'error' | 'warning' | 'info'
  message: string
  suggestion?: string
}

export interface ThemeData {
  name: string
  type: 'dark' | 'light'
  colors: Record<string, string>
  tokenColors: any[]
  semanticTokenColors: Record<string, any>
}

export interface ThemePlugin {
  name: string
  version: string
  description: string
  apply: (theme: ThemeData, config?: any) => ThemeData
  isCompatible: (themeVersion: string) => boolean
}

export interface ThemeSettings {
  variant: string
  contrast: 'low' | 'normal' | 'high'

  syntax: {
    enableSemanticHighlighting: boolean
    bracketPairColorization: boolean
    customLanguageSupport: string[]
  }

  ui: {
    activityBarPosition: 'left' | 'right'
    tabBarStyle: 'classic' | 'modern' | 'minimal'
    statusBarStyle: 'minimal' | 'full'
  }

  customColors?: Record<string, string>
}
