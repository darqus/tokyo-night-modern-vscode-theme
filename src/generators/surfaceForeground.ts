import type { InterfacePalette } from '../types/theme'
import { getContrastRatioAware } from '../core/contrast'
import { pickReadable } from './pickReadable'

// Опциональная настройка интенсивности CodeLens через env
// THEME_CODELENS_INTENSITY: 'auto' | 'primary' | 'muted' | 'subtle'
const CODELENS_INTENSITY = (process.env.THEME_CODELENS_INTENSITY || 'auto')
  .toString()
  .toLowerCase() as 'auto' | 'primary' | 'muted' | 'subtle'

// Автоподбор читаемости для CodeLens: пробуем muted над editor.bg (base),
// если контраст < 4.0 — используем primary
function resolveCodeLens(ip: InterfacePalette): string {
  const bg = ip.bg.base
  if (CODELENS_INTENSITY === 'primary') return ip.textOn.base.primary
  if (CODELENS_INTENSITY === 'muted') {
    // При прямом выборе muted не ломаем намерение, но если совсем плохо,
    // делаем страховку на 3.0 (минимальный порог для muted)
    const r = getContrastRatioAware(ip.textOn.base.muted, bg as any)
    return r >= 3.0 ? ip.textOn.base.muted : ip.textOn.base.primary
  }
  if (CODELENS_INTENSITY === 'subtle') {
    // subtle редко достигает норм, поэтому также страхуемся
    const r = getContrastRatioAware(ip.textOn.base.subtle, bg as any)
    return r >= 3.0 ? ip.textOn.base.subtle : ip.textOn.base.primary
  }
  // auto: предпочитаем muted, если достигает 4.0; иначе primary
  const ratioMuted = getContrastRatioAware(ip.textOn.base.muted, bg as any)
  return ratioMuted >= 4.0 ? ip.textOn.base.muted : ip.textOn.base.primary
}

// Единый список surface-aware foreground токенов → роль текста
// Генератор использует его для присвоения значений, тест — для валидации.
export const SURFACE_FOREGROUND_MAP: Record<
  string,
  (ip: InterfacePalette) => string
> = {
  // Base
  'titleBar.activeForeground': (ip) => ip.textOn.base.primary,
  'titleBar.inactiveForeground': (ip) => ip.textOn.base.muted,
  'commandCenter.foreground': (ip) => ip.textOn.base.muted,
  'commandCenter.activeForeground': (ip) => ip.textOn.base.primary,
  'menubar.selectionForeground': (ip) => ip.textOn.base.primary,
  'editor.foreground': (ip) => ip.textOn.base.primary,
  'editor.findMatchForeground': (ip) => ip.textOn.base.primary,
  'editorWhitespace.foreground': (ip) => ip.textOn.base.subtle,
  'editorCodeLens.foreground': (ip) => resolveCodeLens(ip),
  'activityBar.foreground': (ip) => ip.textOn.base.muted,
  'activityBar.inactiveForeground': (ip) => ip.textOn.base.inactive,
  'sideBar.foreground': (ip) => ip.textOn.base.primary,
  'sideBarTitle.foreground': (ip) => ip.textOn.base.primary,
  'panelTitle.activeForeground': (ip) => ip.textOn.base.primary,
  'panelTitle.inactiveForeground': (ip) => ip.textOn.base.muted,
  'breadcrumb.foreground': (ip) => ip.textOn.base.muted,
  'breadcrumb.focusForeground': (ip) =>
    pickReadable(ip, 'base', 'muted', { mutedMin: 3.0 }),
  'breadcrumb.activeSelectionForeground': (ip) => ip.textOn.base.primary,
  'breadcrumb.hoverForeground': (ip) =>
    pickReadable(ip, 'base', 'muted', { mutedMin: 3.0 }),
  'peekViewResult.fileForeground': (ip) => ip.textOn.base.primary,
  'peekViewResult.lineForeground': (ip) => ip.textOn.base.muted,
  'peekViewResult.selectionForeground': (ip) => ip.textOn.base.primary,
  'terminal.foreground': (ip) => ip.textOn.base.primary,
  'terminal.initialHintForeground': (ip) => ip.textOn.base.muted,
  'settings.headerForeground': (ip) => ip.textOn.base.primary,

  // Elevated
  'menu.foreground': (ip) => ip.textOn.elevated.primary,
  'menu.selectionForeground': (ip) => ip.textOn.elevated.primary,
  'quickInput.foreground': (ip) => ip.textOn.elevated.primary,
  'quickInputList.focusForeground': (ip) => ip.textOn.elevated.primary,
  'sideBarSectionHeader.foreground': (ip) => ip.textOn.elevated.primary,
  'statusBar.foreground': (ip) => ip.textOn.elevated.muted,
  'notifications.foreground': (ip) => ip.textOn.elevated.primary,
  'notificationCenterHeader.foreground': (ip) => ip.textOn.elevated.primary,
  'editorWidget.foreground': (ip) => ip.textOn.elevated.primary,
  'panelSectionHeader.foreground': (ip) => ip.textOn.elevated.primary,
  'keybindingLabel.foreground': (ip) => ip.textOn.elevated.primary,
  'checkbox.foreground': (ip) => ip.textOn.elevated.primary,
  'settings.checkboxForeground': (ip) => ip.textOn.elevated.primary,
  'settings.textInputForeground': (ip) => ip.textOn.elevated.primary,
  'settings.numberInputForeground': (ip) => ip.textOn.elevated.primary,

  // Overlay
  'editorSuggestWidget.foreground': (ip) => ip.textOn.overlay.primary,
  'editorSuggestWidgetStatus.foreground': (ip) => ip.textOn.overlay.muted,
  'peekViewTitleDescription.foreground': (ip) => ip.textOn.overlay.muted,
  'peekViewTitleLabel.foreground': (ip) => ip.textOn.overlay.primary,
}
