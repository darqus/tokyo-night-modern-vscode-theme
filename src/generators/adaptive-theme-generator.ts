import type { ThemeData, ThemeObject } from '../types/theme'
import {
  getAdaptedPalette,
  PaletteModification,
  PaletteVariant,
} from '../palette/adapters'
import { buildColorsWithContext } from '../theme/colors'
import { semanticTokenColors } from '../semanticTokenColors'
import { getTokenColors } from '../tokenColors'

/**
 * Конфигурация темы
 */
export interface ThemeGeneratorConfig {
  name: string
  displayName: string
  variant: PaletteVariant
  customModification?: PaletteModification
  type?: 'dark' | 'light' | 'storm' | 'moon' | 'contrast' | 'pastel'
}

/**
 * Контекст для генерации темы с адаптированной палитрой
 */
export interface ThemeContext {
  adaptedPalette: ReturnType<typeof getAdaptedPalette>
  variant: PaletteVariant
  displayName: string
  type: 'dark' | 'light' | 'storm' | 'moon' | 'contrast' | 'pastel'
}

/**
 * Генератор адаптивных тем
 */
export class AdaptiveThemeGenerator {
  /**
   * Генерирует тему с адаптированной палитрой
   */
  static generateTheme(config: ThemeGeneratorConfig): ThemeObject {
    // Создаем адаптированную палитру
    const adaptedPalette = getAdaptedPalette(config.variant)

    // Определяем тип темы на основе варианта (VS Code поддерживает только dark/light/hc-dark/hc-light)
    let themeType: 'dark' | 'light' = config.type === 'light' ? 'light' : 'dark'
    let vsCodeType: 'dark' | 'light' | 'hc-dark' | 'hc-light' = themeType

    // Проверяем на высококонтрастные темы
    if (
      config.customModification?.contrastBoost &&
      config.customModification.contrastBoost > 1.5
    ) {
      vsCodeType = themeType === 'light' ? 'hc-light' : 'hc-dark'
    }

    // Создаем контекст темы
    const themeContext: ThemeContext = {
      adaptedPalette,
      variant: config.variant,
      displayName: config.displayName,
      type: config.type || 'dark', // Внутренний тип для логики
    }

    // Временно заменяем глобальную extendedPalette
    const originalExtendedPalette =
      require('../palette/extended').extendedPalette
    require('../palette/extended').extendedPalette = adaptedPalette

    try {
      // Генерируем тему с новой палитрой
      const theme: ThemeObject = {
        name: config.name,
        displayName: config.displayName,
        author: 'lod',
        maintainers: ['lod'],
        type: vsCodeType, // Используем валидный VS Code тип
        semanticClass: 'tokyo-night',
        semanticHighlighting: true,
        colors: buildColorsWithContext(themeContext),
        semanticTokenColors: semanticTokenColors,
        tokenColors: getTokenColors(),
      }

      return theme
    } finally {
      // Восстанавливаем оригинальную палитру
      require('../palette/extended').extendedPalette = originalExtendedPalette
    }
  }

  /**
   * Генерирует множественные варианты темы
   */
  static generateThemeVariants(): ThemeObject[] {
    const variants: ThemeGeneratorConfig[] = [
      {
        name: 'tokyo-night-dark',
        displayName: 'Tokyo Night Dark',
        variant: 'tokyo-night',
        type: 'dark',
      },
      {
        name: 'tokyo-night-light',
        displayName: 'Tokyo Night Light',
        variant: 'tokyo-light',
        type: 'light',
      },
      {
        name: 'tokyo-night-storm',
        displayName: 'Tokyo Night Storm',
        variant: 'tokyo-storm',
        type: 'storm',
      },
      {
        name: 'tokyo-night-moon',
        displayName: 'Tokyo Night Moon',
        variant: 'tokyo-night',
        type: 'moon',
      },
      {
        name: 'tokyo-night-high-contrast',
        displayName: 'Tokyo Night High Contrast',
        variant: 'custom',
        customModification: {
          saturationMultiplier: 1.3,
          contrastBoost: 1.5,
          lightnessOffset: 8,
        },
        type: 'contrast',
      },
      {
        name: 'tokyo-night-low-contrast',
        displayName: 'Tokyo Night Low Contrast',
        variant: 'custom',
        customModification: {
          saturationMultiplier: 0.6,
          contrastBoost: 0.5,
          lightnessOffset: -5,
        },
        type: 'dark',
      },
      {
        name: 'tokyo-night-pastel',
        displayName: 'Tokyo Night Pastel',
        variant: 'custom',
        customModification: {
          saturationMultiplier: 0.3,
          lightnessOffset: 20,
          contrastBoost: 0.8,
        },
        type: 'pastel',
      },
    ]

    return variants.map((variant) => this.generateTheme(variant))
  }

  /**
   * Создает кастомную тему
   */
  static createCustomTheme(
    name: string,
    displayName: string,
    modification: PaletteModification,
    type: 'dark' | 'light' | 'storm' | 'moon' | 'contrast' | 'pastel' = 'dark'
  ): ThemeObject {
    return this.generateTheme({
      name,
      displayName,
      variant: 'custom',
      customModification: modification,
      type,
    })
  }

  /**
   * Создает тему из предустановленного варианта палитры
   */
  static createFromPreset(
    name: string,
    displayName: string,
    presetName: keyof typeof paletteVariants,
    type: 'dark' | 'light' | 'storm' | 'moon' | 'contrast' | 'pastel' = 'dark'
  ): ThemeData {
    const preset = paletteVariants[presetName]()

    // Конвертируем палитру в модификацию (упрощенно)
    const modification: PaletteModification = {}

    return this.createCustomTheme(name, displayName, modification, type)
  }
}

/**
 * Утилиты для работы с палитрами
 */
export class PaletteUtils {
  /**
   * Создает градацию между двумя палитрами
   */
  static createGradient(
    baseName: string,
    variant1: PaletteVariant,
    variant2: PaletteVariant,
    steps: number = 5
  ): ThemeObject[] {
    const themes: ThemeObject[] = []

    for (let i = 0; i < steps; i++) {
      const ratio = i / (steps - 1)
      const palette1 = getAdaptedPalette(variant1)
      const palette2 = getAdaptedPalette(variant2)

      // Создаем промежуточные модификации
      const modification: PaletteModification = {
        lightnessOffset: ratio * 10 - 5,
        saturationMultiplier: 0.8 + ratio * 0.4,
      }

      themes.push(
        AdaptiveThemeGenerator.createCustomTheme(
          `${baseName}-${i}`,
          `${baseName} Gradient ${i + 1}`,
          modification
        )
      )
    }

    return themes
  }

  /**
   * Анализирует палитру и предлагает улучшения
   */
  static analyzePalette(palette: ReturnType<typeof getAdaptedPalette>) {
    // Можно добавить анализ контрастности, цветовой гармонии и т.д.
    return {
      dominantHue: 'blue', // Пример
      saturationLevel: 'medium',
      contrastRatio: 4.5,
      accessibility: 'AA',
      colorHarmony: 'analogous',
    }
  }

  /**
   * Экспортирует палитру в различные форматы
   */
  static exportPalette(
    palette: ReturnType<typeof getAdaptedPalette>,
    format: 'json' | 'css' | 'scss' | 'figma'
  ): string {
    switch (format) {
      case 'css':
        return this.exportToCss(palette)
      case 'scss':
        return this.exportToScss(palette)
      case 'figma':
        return this.exportToFigma(palette)
      case 'json':
      default:
        return JSON.stringify(palette, null, 2)
    }
  }

  private static exportToCss(palette: any, prefix = '--tokyo'): string {
    let css = ':root {\n'

    function processObject(obj: any, path: string[] = []): void {
      for (const [key, value] of Object.entries(obj)) {
        if (typeof value === 'string' && value.startsWith('#')) {
          const varName = `${prefix}-${[...path, key].join('-')}`
          css += `  ${varName}: ${value};\n`
        } else if (typeof value === 'object' && value !== null) {
          processObject(value, [...path, key])
        }
      }
    }

    processObject(palette)
    css += '}\n'
    return css
  }

  private static exportToScss(palette: any): string {
    let scss = '// Tokyo Night Palette\n\n'

    function processObject(obj: any, path: string[] = []): void {
      for (const [key, value] of Object.entries(obj)) {
        if (typeof value === 'string' && value.startsWith('#')) {
          const varName = `$tokyo-${[...path, key].join('-')}`
          scss += `${varName}: ${value};\n`
        } else if (typeof value === 'object' && value !== null) {
          processObject(value, [...path, key])
        }
      }
    }

    processObject(palette)
    return scss
  }

  private static exportToFigma(palette: any): string {
    const figmaTokens: any = {}

    function processObject(obj: any, path: string[] = []): void {
      for (const [key, value] of Object.entries(obj)) {
        if (typeof value === 'string' && value.startsWith('#')) {
          const tokenPath = [...path, key]
          let current = figmaTokens

          for (let i = 0; i < tokenPath.length - 1; i++) {
            if (!current[tokenPath[i]]) current[tokenPath[i]] = {}
            current = current[tokenPath[i]]
          }

          current[tokenPath[tokenPath.length - 1]] = {
            value: value,
            type: 'color',
          }
        } else if (typeof value === 'object' && value !== null) {
          processObject(value, [...path, key])
        }
      }
    }

    processObject(palette)
    return JSON.stringify(figmaTokens, null, 2)
  }
}
