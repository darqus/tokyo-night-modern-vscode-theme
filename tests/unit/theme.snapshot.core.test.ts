import { generateTheme, loadEnvVars } from '../../src/generators/theme'

/**
 * Partial snapshot: Core foundational UI (Base Colors, Editor, Sidebar, Activity Bar, Status Bar, Tabs, Title Bar, Editor Groups)
 * Purpose: Smaller diff surface when adjusting foundational surfaces without noise from peripheral features.
 */

describe('Theme partial snapshot: core surfaces', () => {
  it('should match core surface tokens snapshot', () => {
    const theme = generateTheme(loadEnvVars())
    const coreTokenPrefixes = [
      'foreground',
      'icon.',
      'descriptionForeground',
      'disabledForeground',
      'focusBorder',
      'errorForeground',
      'selection.background',
      'editor.',
      'sideBar',
      'activityBar',
      'statusBar',
      'statusBarItem.',
      'tab.',
      'titleBar',
      'editorGroup',
    ]

    const subset: Record<string, string> = {}
    for (const [k, v] of Object.entries(theme.colors)) {
      if (coreTokenPrefixes.some((p) => k.startsWith(p))) subset[k] = v
    }

    expect(subset).toMatchSnapshot()
  })
})
