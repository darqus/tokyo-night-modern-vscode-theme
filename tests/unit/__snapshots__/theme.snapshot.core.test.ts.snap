// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Theme partial snapshot: core surfaces should match core surface tokens snapshot 1`] = `
{
  "activityBar.activeBackground": "#1a1b26",
  "activityBar.activeBorder": "#7dcfff",
  "activityBar.background": "#1f2230",
  "activityBar.foreground": "#99a3cc",
  "activityBar.inactiveForeground": "#8087a7",
  "activityBarBadge.background": "#007acc",
  "activityBarBadge.foreground": "#ffffff",
  "descriptionForeground": "#99a3cc",
  "disabledForeground": "#99a3cc",
  "editor.background": "#1a1b26",
  "editor.findMatchBackground": "#33b2ff38",
  "editor.findMatchBorder": "#29acf899",
  "editor.findMatchForeground": "#afb9e2",
  "editor.findMatchHighlightBackground": "#36b5ff29",
  "editor.focusedStackFrameHighlightBackground": "#88b0ff42",
  "editor.foreground": "#afb9e2",
  "editor.hoverHighlightBackground": "#81aaff1f",
  "editor.inactiveSelectionBackground": "#81aaff1f",
  "editor.lineHighlightBackground": "#81aaff1f",
  "editor.rangeHighlightBackground": "#81aaff1f",
  "editor.selectionBackground": "#88b0ff42",
  "editor.stackFrameHighlightBackground": "#81aaff1f",
  "editor.wordHighlightBackground": "#81aaff1f",
  "editor.wordHighlightStrongBackground": "#88b0ff42",
  "editorGroup.border": "#202230",
  "editorGroup.dropBackground": "#23273933",
  "editorGroupHeader.noTabsBackground": "#1f2230",
  "editorGroupHeader.tabsBackground": "#1f2230",
  "editorGroupHeader.tabsBorder": "#202230",
  "errorForeground": "#f38095",
  "focusBorder": "#00000000",
  "foreground": "#afb9e2",
  "icon.foreground": "#8991b3",
  "selection.background": "#88b0ff42",
  "sideBar.background": "#181923",
  "sideBar.border": "#202230",
  "sideBar.foreground": "#afb9e2",
  "sideBarSectionHeader.background": "#232739",
  "sideBarSectionHeader.foreground": "#afb9e2",
  "sideBarTitle.foreground": "#afb9e2",
  "statusBar.background": "#1f2230",
  "statusBar.border": "#202230",
  "statusBar.foreground": "#99a3cc",
  "statusBarItem.compactHoverBackground": "#81aaff1f",
  "statusBarItem.hoverBackground": "#81aaff1f",
  "statusBarItem.hoverForeground": "#afb9e2",
  "statusBarItem.prominentBackground": "#007acc",
  "statusBarItem.prominentForeground": "#ffffff",
  "statusBarItem.prominentHoverBackground": "#1f8ad2",
  "statusBarItem.prominentHoverForeground": "#ffffff",
  "statusBarItem.remoteBackground": "#007acc",
  "statusBarItem.remoteForeground": "#ffffff",
  "statusBarItem.remoteHoverBackground": "#1f8ad2",
  "statusBarItem.remoteHoverForeground": "#ffffff",
  "tab.activeBackground": "#1a1b26",
  "tab.activeBorderTop": "#7dcfff",
  "tab.activeForeground": "#afb9e2",
  "tab.hoverBackground": "#81aaff1f",
  "tab.hoverForeground": "#afb9e2",
  "tab.inactiveBackground": "#1f2230",
  "tab.inactiveForeground": "#99a3cc",
  "titleBar.activeBackground": "#1f2230",
  "titleBar.activeForeground": "#8991b3",
  "titleBar.border": "#202230",
  "titleBar.inactiveBackground": "#1f2230",
  "titleBar.inactiveForeground": "#8087a7",
}
`;
